[mlir][ArmSME] Make use of backend function attributes for enabling ZA storage

MacDue opened the pull request:
Previously, we were inserting za.enable/disable intrinsics for functions with the "arm_za" attribute (at the MLIR level), rather than using the backend attributes. This was done to avoid a dependency on the SME ABI functions from compiler-rt (which have only recently been implemented).

Doing things this way did have correctness issues, for example, calling a streaming-mode function from another streaming-mode function (both with ZA enabled) would lead to ZA being disabled after returning to the caller (where it should still be enabled). Fixing issues like this would require re-doing the ABI work already done in the backend within MLIR.

Instead, this patch switches to use the "arm_new_za" (backend) attribute for enabling ZA for an MLIR function. For the integration tests, this requires some way of linking the SME ABI functions. This is done via the `%arm_sme_abi_shlib` lit substitution. By default, this expands to a stub implementation of the SME ABI functions, but this can be overridden by providing the `ARM_SME_ABI_ROUTINES_SHLIB` CMake cache variable (pointing it at an alternative implementation). For now, the ArmSME integration tests pass with just stubs, as we don't make use of nested ZA-enabled calls.

A future patch may add an option to compiler-rt to build the SME builtins into a standalone shared library to allow easily building/testing with the actual implementation.

MacDue committed:
[mlir][ArmSME] Make use of backend function attributes for enabling ZA storage

Previously, we were inserting za.enable/disable intrinsics for functions
with the "arm_za" attribute (at the MLIR level), rather than using the
backend attributes. This was done to avoid a dependency on the SME ABI
functions from compiler-rt (which have only recently been implemented).

Doing things this way did have correctness issues, for example, calling
a streaming-mode function from another streaming-mode function (both
with ZA enabled) would lead to ZA being disabled after returning to the
caller (where it should still be enabled). Fixing issues like this would
require re-doing the ABI work already done in the backend within MLIR.

Instead, this patch switches to use the "arm_new_za" (backend) attribute
for enabling ZA for a MLIR function. For the integration tests this
requires some way of linking the SME ABI functions. This has been done
by adding a mlir_arm_sme_runtime library, which includes the
implementation from compiler-rt, which can then be linked via the
`-shared-libs` flag.

To build the mlir_arm_sme_runtime the target has to be AArch64, and
the host compiler must be able to assemble SME instructions (this is
supported in recent versions of clang). Note that the host being AArch64
is already assumed by the integration tests linking other runtime
libraries (e.g. mlir_c_runner_utils).

MacDue commented:
cc @c-rhodes, @banach-space 

Keeping this as a draft for now as I'm a little unhappy with the current solution for the `mlir_arm_sme_runtime` library.

c-rhodes commented on mlir/lib/ExecutionEngine/CMakeLists.txt:
```diff
@@ -177,6 +178,39 @@ if(LLVM_ENABLE_PIC)
+      message(FATAL_ERROR "Host compiler must be able to assemble AArch64 SME instructions to build the ArmSME runtime.")
+    endif()
+
+    # FIXME: This is very far from ideal, but enabling compiler-rt in the main
```
I managed to build a minimal (aarch64 builtins only) standalone compiler-rt on aarch64 with the following:
```
# Get recent clang with SME support. This is the most recent but older versions (within reason) should work as well.
wget https://github.com/llvm/llvm-project/releases/download/llvmorg-17.0.4/clang+llvm-17.0.4-aarch64-linux-gnu.tar.xz
tar xvf clang+llvm-17.0.4-aarch64-linux-gnu.tar.xz

cd <path-to-llvm-project>

cmake -G Ninja \
  -S compiler-rt -B build-compiler-rt \
  -DCMAKE_C_COMPILER=<path-to>/clang+llvm-17.0.4-aarch64-linux-gnu/bin/clang-17 \
  -DCMAKE_CXX_COMPILER=<path-to>/clang+llvm-17.0.4-aarch64-linux-gnu/bin/clang++ \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_INSTALL_PREFIX=<install-path> \
  -DLLVM_CMAKE_DIR=cmake \
  -DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON \
  -DCMAKE_C_COMPILER_TARGET=aarch64-linux-gnu \
  -DCOMPILER_RT_BUILD_GWP_ASAN=OFF \
  -DCOMPILER_RT_BUILD_LIBFUZZER=OFF \
  -DCOMPILER_RT_BUILD_MEMPROF=OFF \
  -DCOMPILER_RT_BUILD_ORC=OFF \
  -DCOMPILER_RT_BUILD_PROFILE=OFF \
  -DCOMPILER_RT_BUILD_SANITIZERS=OFF \
  -DCOMPILER_RT_BUILD_XRAY=OFF

ninja -C build-compiler-rt install

tree <install-path>
install
└── lib
    └── linux
        ├── clang_rt.crtbegin-aarch64.o
        ├── clang_rt.crtend-aarch64.o
        └── libclang_rt.builtins-aarch64.a
```

As far as I can tell there's three approaches here:

1. Build compiler-rt together with llvm+clang+mlir.
2. Build standalone compiler-rt separately (as above).
3. Build minimal SME runtime from sources when building llvm+mlir (as you're doing here).

All options require recent host toolchain with SME support.

Options 1 and 3 the runtime is integrated into normal build. I guess for both these options it wouldn't be necessary to add a CMake flag to specify the path to the runtime. Option 2 the runtime is built separately and a CMake flag would be necessary.

Options 1 and 3 may require updating existing build to update host toolchain if current doesn't support SME. Option 2 existing builds wouldn't need to change, except for adding the CMake flag with path to runtime lib.

Option 1 requires building Clang. I like this option the least, it requires building much more than we need as you say.

Comparing options 2 and 3. Option 2 requires fewer changes to MLIR, most (all?) of the changes in this CMake file could be removed? Option 2 would also be insulated from changes to the SME runtime implementation in compiler-rt if it were to ever change.

Food for thought. I'm gravitating towards Option 2, let us know if I've missed anything or what you think.

MacDue replied:
Note that `compiler-rt` only builds the builtin as a static library, so we can't load those at runtime in the MLIR tests. 
```
      add_compiler_rt_runtime(clang_rt.builtins
                              STATIC
                              ARCHS ${arch}
                              DEPS ${deps_${arch}}
                              SOURCES ${${arch}_SOURCES}
                              DEFS ${BUILTIN_DEFS}
                              CFLAGS ${BUILTIN_CFLAGS_${arch}}
                              PARENT_TARGET builtins)
```

MacDue replied:
So, I think option 2 likely requires changing compiler-rt to allow building the builtins into a shared library (probably just the SME builtins).  This would be via some flag on the build like `COMPILER_RT_BUILD_STANDALONE_AARCH64_SME_RUNTIME`.

MacDue replied:
I'm not sure how option 1 would work, since by default `compiler-rt` won't build a shared library we can use, and the `LLVM_ENABLE_RUNTIMES="compiler-rt"` does not seem to expose options from compiler-rt into the main build.

c-rhodes replied:
> Note that `compiler-rt` only builds the builtins as a static library, so we can't load those at runtime in the MLIR tests.

Ah, I didn't consider that. Surprised that's not configurable, I wonder if there's a reason for only building as static.

MacDue replied:
I think these libraries are intended to be linked into your binary by clang, so in that case it makes sense for them to be static. There is an option to build parts into a shared library, for example:
```
option(COMPILER_RT_BUILD_STANDALONE_LIBATOMIC
  "Build standalone shared atomic library."
  OFF)
```
Which is what gave me the idea for the `COMPILER_RT_BUILD_STANDALONE_AARCH64_SME_RUNTIME` option.

banach-space replied:
> Options 1 and 3 may require updating existing build to update host toolchain if current doesn't support SME. 

IIUC, all options will require a "recent" toolchain, right?

My view is that:
* Option 1 would be the cleanest/simplest and would work fine for pre-commit/post-commit testing (as Buildbots tend to build everything anyway). However, this would be too painful for developers.
* Option 2 would force us to use 2-stage builds for testing. This would be more work for developers and potentially extra set-up for buildbots.
* Options 3 is the simplest from the developer's perspective. And that's what I'd personally optimise for. I do wonder though - wouldn't it be possible to add this target in `compiler-rt` instead?

c-rhodes replied:
> > Options 1 and 3 may require updating existing build to update host toolchain if current doesn't support SME.
> 
> IIUC, all options will require a "recent" toolchain, right?

Yeah, but for option 2 that toolchain wouldn't need to be used to build llvm+mlir, only to build compiler-rt as standalone.

> My view is that:
> 
>     * Option 1 would be the cleanest/simplest and would work fine for pre-commit/post-commit testing (as Buildbots tend to build everything anyway). However, this would be too painful for developers.

As Ben pointed out, this is currently possible since compiler-rt builds static libs.

banach-space replied:
> Yeah, but for option 2 that toolchain wouldn't need to be used to build llvm+mlir, only to build compiler-rt as standalone.

Ah, thanks! IMHO, this sort of things tend to be tricky to document, i.e. "use whatever toolchain for A, and then something special for B". It can be done, just a bit tricky.

> As Ben pointed out, this is currently possible since compiler-rt builds static libs.

Yes, but we shouldn't assume that that can't be updated, right? (also - did you mean _not_ possible?)

c-rhodes replied:
>Yes, but we shouldn't assume that that can't be updated, right? (also - did you mean not possible?)

Of course, and yeah I meant isn't possible :)

banach-space commented on mlir/lib/ExecutionEngine/CMakeLists.txt:
```diff
@@ -177,6 +178,39 @@ if(LLVM_ENABLE_PIC)
+    endif()
+
+    if (NOT DEFINED MLIR_ARM_SME__CAN_ASSEMBLE_ARM_SME)
+      # This should work on an AArch64 host with a recent version of clang.
```
[nit] Rather than "recent" (which is a relative term), use something like "Clang 17 or higher". This way people will know exactly what's required.

MacDue replied:
Decided not to go with this approach last week. Instead, now within MLIR only stubs for the ABI routines are build (which is enough for our current set of tests to pass, as we don't rely on ZA-enabled functions calling each other). The use of stubs can be overridden by setting `ARM_SME_ABI_ROUTINES_SHLIB` to a path an alternate implementation. 

A later patch could be made to compiler-rt to allow building the SME RT standalone (which would allow us to make use of the actual functions fairly easily).

MacDue committed:
Default to ABI stubs configurable via ARM_SME_ABI_ROUTINES_SHLIB

This removes the direct dependency on compiler-rt and instead includes
ABI stub routines in MLIR. Our current tests pass with only stubs, as
we're not making nested ZA-enabled calls. Using these stubs can be
overridden by setting the ARM_SME_ABI_ROUTINES_SHLIB CMake cache
variable to a path to an alternate implementation.

MacDue committed:
Fixup

Dinistro commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -1583,7 +1583,8 @@ static void processPassthroughAttrs(llvm::Function *func, LLVMFuncOp funcOp) {
     // explicit attribute.
     // Also skip the vscale_range, it is also an explicit attribute.
```
Nit: Can you adapt the comment to be either more general or explicitly list this newly added attibute?

MacDue replied:
I've reworked this to a `ExplicitAttributes` list rather than this ever-growing 'or' expression :+1:

c-rhodes commented on mlir/lib/ExecutionEngine/ArmSMEStub.cpp:
nit: ArmSMEStubs.cpp or ArmSMERuntimeStubs.cpp?

c-rhodes commented on mlir/test/CMakeLists.txt:
```diff
@@ -139,6 +141,10 @@ if(MLIR_ENABLE_ROCM_RUNNER)
   list(APPEND MLIR_TEST_DEPENDS mlir_rocm_runtime)
 endif()
 
+if (MLIR_RUN_ARM_SME_TESTS)
+  list(APPEND MLIR_TEST_DEPENDS mlir_arm_sme_abi_stubs)
+endif()
```
unless `ARM_SME_ABI_ROUTINES_SHLIB` is set?

c-rhodes commented on mlir/test/lit.cfg.py:
```diff
@@ -126,6 +131,15 @@ def add_runtime(name):
+if config.mlir_run_arm_sme_tests:
+    config.substitutions.append(
+        (
+            "%arm_sme_abi_shlib",
```
nit: `%arm_sme_runtime` or `%arm_sme_runtime_shlib`?

MacDue replied:
I think "runtime" implies these are something more substantial, but this library should only ever contain a few SME ABI builtins.

c-rhodes replied:
fair enough, that's what these are tho, runtime support routines [1]:
> Every platform that supports SME must provide the following runtime support routines:

[1] https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#sme-support-routines

banach-space replied:
I also think that "runtime" can be interpreted as something quite substantial. Having said that, I agree with Cullen that we should strive for consistency with the spec. I suggest "arm_sme_support_routines_shlib".

In any case, I wouldn't worry about this particular name too much :) (we can always rename later)

c-rhodes reviewed (approved):
I've left some minor nits, but otherwise LGTM, cheers!

banach-space commented on mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.h:
```diff
@@ -24,15 +24,19 @@ namespace arm_sme {
 // In a locally streaming function PSTATE.SM is kept internal and the callee
 // manages it on entry/exit.
-enum class ArmStreaming { Default = 0, Locally = 1 };
+enum class ArmStreamingMode { Default = 0, Locally = 1 };
```
Wondering whether this shouldn't be something like:
```cpp
enum class ArmStreamingMode { Disabled = 0, ArmStreaming, ArmLocallyStreaming = 1 };
```
Basically, we'd want this to be consistent with ACLE (which doesn't have `Default`) as well as `ArmZaMode` (which defines `Disabled`).

MacDue replied:
Make sense :+1:
I've removed the "Default" state and added a "Disabled" state. The old "Default" is now"Streaming".

banach-space replied:
Hm, I still see `Default`?

MacDue committed:
Fixups

- ArmSMEStub.cpp -> ArmSMEStubs.cpp
- Move enums to tablegen (to get generated stringification)
- Make enums closer to ACLE
  * Remove "Default" mode
  * Similar naming:
  * ArmZaMode::New -> ArmZaMode::NewZA
  * ArmStreamingMode::Locally -> ArmStreamingMode::StreamingLocally

MacDue committed:
Fix copy/paste error :)

banach-space reviewed (approved):
LGTM, thanks!

MacDue committed:
Update recently added tests

These required a few little changes to avoid nested ZA enabled calls.

MacDue commented:
@c-rhodes I had to update a few recent tests to avoid nested ZA-enabled calls, which only works when using the real ABI runtime. If the subs are used the backend emitted checks end up using some invalid state. See: https://github.com/llvm/llvm-project/pull/71044/commits/84c9d7fcaff53ca08125d75f5f196c15d69ee751

c-rhodes commented on mlir/lib/Dialect/ArmSME/Transforms/EnableArmStreaming.cpp:
```diff
@@ -48,46 +49,39 @@ namespace arm_sme {
-    getOperation()->setAttr(attr, UnitAttr::get(&getContext()));
+    auto unitAttr = UnitAttr::get(&getContext());
+
+    if (streamingMode == ArmStreamingMode::Disabled)
+      return;
```
nit: move above `unitAttr` or incorporate into top if

c-rhodes commented on mlir/test/Dialect/ArmSME/enable-arm-streaming.mlir:
```diff
@@ -1,13 +1,13 @@
 // RUN: mlir-opt %s -enable-arm-streaming -verify-diagnostics | FileCheck %s
-// RUN: mlir-opt %s -enable-arm-streaming=mode=locally -verify-diagnostics | FileCheck %s -check-prefix=CHECK-LOCALLY
-// RUN: mlir-opt %s -enable-arm-streaming=enable-za -verify-diagnostics | FileCheck %s -check-prefix=CHECK-ENABLE-ZA
+// RUN: mlir-opt %s -enable-arm-streaming=streaming-mode=streaming-locally  -verify-diagnostics | FileCheck %s -check-prefix=CHECK-LOCALLY
```
nit: empty space
```suggestion
// RUN: mlir-opt %s -enable-arm-streaming=streaming-mode=streaming-locally -verify-diagnostics | FileCheck %s -check-prefix=CHECK-LOCALLY
```

c-rhodes commented:
> @c-rhodes I had to update a few recent tests to avoid nested ZA-enabled calls, which only works when using the real ABI runtime. If the subs are used the backend emitted checks end up using some invalid state. See: [84c9d7f](https://github.com/llvm/llvm-project/commit/84c9d7fcaff53ca08125d75f5f196c15d69ee751)

Thanks for heads up. I left a couple of minor nits I spotted but otherwise still LGTM, cheers.

MacDue committed:
Fixups
